### PR TITLE
KTIJ-4841 Add "Introduce Constant" to file scope for kotlin

### PR DESCRIPTION
**Pull request by XiangRongLin**
partly sovles https://youtrack.jetbrains.com/issue/KTIJ-4841

1. I oriented myself heavily on "introduce property"
2. This only support extracting to top level/file scope
3. Only `const val` can be extracted, because otherwise "introduce property" can be used
4. I hopefully used the visitor correctly to determine whether an element is constant. I got that tip here https://intellij-support.jetbrains.com/hc/en-us/community/posts/4404981090706-How-to-check-if-PsiElement-is-a-kotlin-constant

### CODE REVIEW

File: plugins/kotlin/idea/tests/testData/refactoring/introduceConstant/binaryExpression/stringAddition.kt.after [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// EXTRACTION_TARGET: property with initializer
+val a = 1
+
+private const val s = "1" + "1"
+
+fun foo(): String {
+    val x = 1+1
```
Comment (TWiStErRob) [inline plugins/kotlin/idea/tests/testData/refactoring/introduceConstant/binaryExpression/stringAddition.kt.after:7]: Does this test actually verify that the unrelated code is reformatted? Or only the actual programmatic elements become diffed?

Comment (XiangRongLin) [reply plugins/kotlin/idea/tests/testData/refactoring/introduceConstant/binaryExpression/stringAddition.kt.after:7]: This comes from me orienting myself on the other tests, see `extractTrue` or `extractIntegerLiteral` which I copied from introduce property

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            continuation
+        )
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        throw AssertionError("$INTRODUCE_CONSTANT can only be invoked from editor")
+    }
+}
+
+val INTRODUCE_CONSTANT: String = KotlinBundle.message("introduce.constant")
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt:173]: pls be aware of `@Nls`:
```
    val INTRODUCE_CONSTANT
        @Nls
        get() = KotlinBundle.message("introduce.constant")
```

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                    || this is KtLiteralStringTemplateEntry
+        }
+    }
+
+    private fun getNameSuggestions(property: KtProperty): List<String> {
+        val initializerValue = (property.initializer as? KtStringTemplateExpression)?.plainContent
+        val identifierValue = property.identifyingElement?.text
+
+        return listOfNotNull(initializerValue, identifierValue).map { NameUtil.capitalizeAndUnderscore(it) }
+
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt:139]: empty line ?

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                    super.visitKtElement(element)
+                    if (element !is KtOperationReferenceExpression && element !is KtBinaryExpression) {
+                        isConst = isConst && element.isConst()
+                    }
+                }
+            }.visitBinaryExpression(this)
+
+            return isConst
+        } else {
+            return this is KtConstantExpression
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt:128]: not sure about whitespaces / new lines

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    private fun PsiElement.isNotConst() = !isConst()
+
+    private fun PsiElement.isConst(): Boolean {
+        if (this is KtBinaryExpression) {
+            var isConst = true
+            object : KotlinRecursiveElementVisitor() {
+                override fun visitKtElement(element: KtElement) {
+                    super.visitKtElement(element)
```
[the remark this answers is not in the document]
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: Added an if check around it

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                    if (element !is KtOperationReferenceExpression && element !is KtBinaryExpression) {
```
[the remark this answers is not in the document]
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: I'll take a loot at it later.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt (hunk 4 above)
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt:132]: I put each type on a seperate line because otherwise the line would be very long. I could do 
```kotlin
        } else {
            return this is KtConstantExpression || this is KtLiteralStringTemplateEntry
                   || (this is KtStringTemplateExpression && !this.hasInterpolation())
        }
```

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+            return isConst
+        } else {
+            return this is KtConstantExpression
+                    || (this is KtStringTemplateExpression && !this.hasInterpolation())
+                    || this is KtLiteralStringTemplateEntry
+        }
+    }
+
+    private fun getNameSuggestions(property: KtProperty): List<String> {
+        val initializerValue = (property.initializer as? KtStringTemplateExpression)?.plainContent
```
[the remark this answers is not in the document]
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: done

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt (hunk 3 above)
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: removed

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt (hunk 2 above)
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: Not too sure I completly understood the doc of it, but i added it.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: I added a field which contains all operators which are constant and used that to check like in `KtOperationReferenceExpression` Because the file grew a bit, I moved the `isConst` logic into a seperate util file

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt (hunk 4 above)
Comment (TWiStErRob) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt]: I would put the 3 distinct logical conditions on separate lines.

Alternatively (not sure if this is in line with the IJ codebase) I prefer to be explicit when doing complex Boolean expressions:
```kt
return when (this) {
    is KtConstantExpression -> true
    is KtLiteralStringTemplateEntry -> true
    is KtStringTemplateExpression -> !this.hasInterpolation()
    else -> false
}
```
This has the added benefit of having a nice place to put comments on each case explaining them. And the code after the arrows have place to grow.

File: plugins/kotlin/idea/tests/testData/refactoring/introduceConstant/binaryExpression/integerPlusPartNotConst.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+// EXTRACTION_TARGET: property with initializer
+val a = 1
+
+fun foo(): Int {
+    return<selection>1+getNumber()</selection>
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/idea/tests/testData/refactoring/introduceConstant/binaryExpression/integerPlusPartNotConst.kt:5]: 💯

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        is KtPrefixExpression -> {
+            val visitor = ConstKotlinRecursiveElementVisitor(root = this)
+            visitor.visitPrefixExpression(this)
+            return visitor.isConst
+        }
+        else -> {
+            return this is KtConstantExpression
+                    || (this is KtStringTemplateExpression && !this.hasInterpolation())
+                    || this is KtLiteralStringTemplateEntry
+                    || (this is KtOperationReferenceExpression && constOperatorNames.contains(this.resolveToCall()?.resultingDescriptor?.fqNameOrNull()))
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt:66]: I think we could support `KtParenthesizedExpression` as well to handle expressions like `(2+2)*2`.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import org.jetbrains.kotlin.idea.structuralsearch.visitor.KotlinRecursiveElementVisitor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+
+/**
+ * [FqName] of operators which can be used in a kotlin const expression.
+ */
+private val constOperatorNames = listOf("Byte", "Short", "Int", "Long", "UByte", "UShort", "UInt", "ULong")
+    .flatMap {
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt:15]: As well there are many others like binary operators `shl` (e.g. `1 shl 10` that is 1024), `and`, `or` etc - that is quite often used as constants

**Comment by vladimirdolzhenko** [comment]
Usually you don't need to force push your changes to your feature branch as it will be squashed and merged as a single (usually) commit.

Force push makes review a bit harder. (But I see you've done lots of work that makes sense to force push it)

**Comment by vladimirdolzhenko** [comment]
I gonna check compiler internals on how do it checks iff expression could be handled as `const` (e.g. within companion object)

**Comment by vladimirdolzhenko** [comment]
Sorry for misleading you with visitors - there is way more useful approach already performed by FrontEnd of compiler:

https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/resolve/constants/evaluate/ConstantExpressionEvaluator.kt#L350

basically, bindindContext has a flag iff the expression is a compile time constant.

**Comment by XiangRongLin** [comment]
I updated the code with your suggestion of using `ConstantExpressionEvaluator`.

For bindingContext I used `analyze(BodyResolveMode.PARTIAL)`, because it seems what most other usages of `ConstantExpressionEvaluator` also used. I myself don't quite understand that parameter.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt (hunk 9 above)
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt]: Automatically included by using `ConstantExpressionEvaluator`.
I added a test for this

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt (hunk 10 above)
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/IntroduceConstUtil.kt]: Automatically included by using `ConstantExpressionEvaluator`.
I added a test for this

**Comment by XiangRongLin** [comment]
The analysis shows as failing, but when i go to the details it shows as succesfull. Can it be retriggered manually?

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        val adjustedElements = (elements.singleOrNull() as? KtBlockExpression)?.statements ?: elements
+        when {
+            adjustedElements.isEmpty() -> {
+                showErrorHintByKey(
+                    project, editor, "cannot.refactor.no.expression",
+                    INTRODUCE_CONSTANT
+                )
+            }
+            else -> {
+                val options = ExtractionOptions(extractAsProperty = true)
```
[the remark this answers is not in the document]
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt:74]: I left out the option of extration to companion object, because like you said it's more a matter of `ExtractionEngine`. 

I tried a bit unsuccessfully to implement it, but found that extracting to companion object in itself is another very complex topic, which I did not want to do here.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                val extractionData = ExtractionData(file, adjustedElements.toRange(), target, null, options)
+                ExtractionEngine(helper).run(editor, extractionData) {
```
[the remark this answers is not in the document]
Comment (XiangRongLin) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt:76]: I tried around a bit with `project.executeCommand` with `runWriteAction` inside, but I did not manage to get the wanted behaviour. The code in `KotlinIntroduceVariableHandler´ looks completly different from `KotlinIntroducePropertyHandler`, so I don't know what exactly I have to do.

`KotlinIntroducePropertyHandler` (CTRL + ALT + F) has the same issue, since I oriented myself here

**Comment by vladimirdolzhenko** [comment]
Sorry for the long pause. There were some minor issues (refactoring actions are registered now in `refactorings.xml`, `after` test file names has to be exactly the same as original one while `IntegerCompareTo.kt.after` has a capital letter in the beginning, original file is `integerCompareTo.kt` - most probably you are on case-insensitive FS like MacOS or Windows) that I fixed.

The problems related to undos will be covered in a dedicated ticket.

PR is merged as https://github.com/JetBrains/intellij-community/commit/0928661ce5d080d866db8fc5bb65135fe37091a9

Thanks a lot for the contribution.

**Comment by XiangRongLin** [comment]
Yay 🎉 
No worries about the pause. I'm on the other side of PR often enough, so I know that they can happen in Open Source.

**Comment by XiangRongLin** [comment]
By they way, is there a reason why this commit is not showing on my profile or the contributors page? https://github.com/JetBrains/intellij-community/graphs/contributors

**Comment by vladimirdolzhenko** [comment]
it shows only top-100 contributors, e.g. @artfly who is #100 has 369 commits
